polygeist/tools/cgeist/Lib/Attributes.cc

```diff
@@ -7,166 +7,388 @@
+  return Attrs.getNamed(AttrName);
 }
 
 Optional<NamedAttribute>
-AttrBuilder::getAttr(llvm::Attribute::AttrKind kind) const {
-  StringRef attrName = llvm::Attribute::getNameFromAttrKind(kind);
-  return getAttr(attrName);
+AttrBuilder::getAttr(llvm::Attribute::AttrKind Kind) const {
+  StringRef AttrName = llvm::Attribute::getNameFromAttrKind(Kind);
+  return getAttr(AttrName);
+}
+
+StringAttr AttrBuilder::createStringAttr(Twine AttrName,
+                                         Optional<StringLiteral> Prefix,
+                                         MLIRContext &Ctx) {
+  return (Prefix) ? StringAttr::get(&Ctx, *Prefix + "." + AttrName)
+                  : StringAttr::get(&Ctx, AttrName);
+}
+
+AttrBuilder &AttrBuilder::addAttributeImpl(llvm::Attribute::AttrKind Kind,
+                                           Optional<StringLiteral> Dialect,
+                                           AddAttrFuncPtr AddAttrPtr) {
+  assert(AddAttrPtr && "'AddAttrPtr' should be a valid function pointer");
+
+  // TODO: Replace with std::invoke once C++17 headers are available.
+  auto Invoke = [this](AddAttrFuncPtr AddAttrPtr,
+                       auto... Args) -> AttrBuilder & {
+    return (this->*AddAttrPtr)(Args...);
+  };
+
+  OpBuilder Builder(&Ctx);
+  StringRef AttrName = llvm::Attribute::getNameFromAttrKind(Kind);
+  NamedAttribute NamedAttr(createStringAttr(AttrName, Dialect, Ctx),
+                           Builder.getUnitAttr());
+  return Invoke(AddAttrPtr, NamedAttr);
+}
+
+AttrBuilder &AttrBuilder::addAttributeImpl(llvm::Attribute::AttrKind Kind,
+                                           mlir::Type Ty,
+                                           Optional<StringLiteral> Dialect,
+                                           AddAttrFuncPtr AddAttrPtr) {
+  assert(AddAttrPtr && "'AddAttrPtr' should be a valid function pointer");
+
+  // TODO: Replace with std::invoke once C++17 headers are available.
+  auto Invoke = [this](AddAttrFuncPtr AddAttrPtr,
+                       auto... Args) -> AttrBuilder & {
+    return (this->*AddAttrPtr)(Args...);
+  };
+
+  OpBuilder Builder(&Ctx);
+  StringRef AttrName = llvm::Attribute::getNameFromAttrKind(Kind);
+  NamedAttribute NamedAttr(createStringAttr(AttrName, Dialect, Ctx),
+                           mlir::TypeAttr::get(Ty));
+  return Invoke(AddAttrPtr, NamedAttr);
+}
+
+AttrBuilder &
+AttrBuilder::addAttributeImpl(llvm::Attribute::AttrKind Kind, uint64_t Val,
+                              AddRawIntAttrFuncPtr AddRawIntAttrPtr) {
+  assert(AddRawIntAttrPtr &&
+         "'AddRawIntAttrPtr' should be a valid function pointer");
+
+  // TODO: Replace with std::invoke once C++17 headers are available.
+  auto Invoke = [this](AddRawIntAttrFuncPtr AddRawIntAttrPtr,
+                       auto... Args) -> AttrBuilder & {
+    return (this->*AddRawIntAttrPtr)(Args...);
+  };
+
+  switch (Kind) {
+  case llvm::Attribute::AttrKind::Alignment:
+  case llvm::Attribute::AttrKind::StackAlignment:
```

Comment:
The maximum stack alignment is not `llvm::Value::MaximumAlignment`, but `0x100`

Reply:
done